### PR TITLE
Pin, Unpin, PinBox

### PR DESCRIPTION
**Pull request by withoutboats**
Implementing rust-lang/rfcs#2349 (do not merge until RFC is merged)

@bors r? @cramertj

### CODE REVIEW

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -896,3 +896,96 @@ impl<T> Generator for Box<T>
...
+    /// moved out of this reference.
+    pub unsafe fn get_mut<'a>(this: &'a mut PinBox<T>) -> &'a mut T {
+        &mut *this.inner
+    }
+
+    /// Convert this PinBox into an unpinned Box.
+    ///
+    /// This function is unsafe. Users must guarantee that the data is never
+    /// moved out of the box.
+    pub unsafe fn unpin(this: PinBox<T>) -> Box<T> {
```
Comment (cramertj) [inline src/liballoc/boxed.rs]: It seems like there should also be an `unpin_safe` for `T: Unpin` (all of the other functions can go through `Pin` for the `Unpin`-specific behavior, but this one can't).

**Comment by cramertj** [comment]
This looks great to me! s/`issue = "0"`/`issue = "<tracking issue #>"` once a tracking issue is opened, but r=me otherwise.

**Comment by withoutboats** [comment]
htmldoc failure, no idea how it could be related to this change

**Comment by kennytm** [comment]
Blocked by #49123.

**Comment by withoutboats** [comment]
I can change the order of the bounds and it should pass?

**Comment by withoutboats** [comment]
@kennytm also p=100 was because futures 0.2 is blocked on this landing in nightly (i assume thats what your confused face was about)

**Comment by kennytm** [comment]
@withoutboats if you need to change the test I suggest you `// ignore` it, because changing the order may break future PRs instead. 

I'd like to let #49091 get merged first, since it fixes a very commonly encountered ICE.

**Comment by alexcrichton** [comment]
@bors: retry

sorry bors, this already failed

**Pull request merged**